hypothesis-python/src/hypothesis/internal/compat.py

```diff
@@ -188,3 +191,68 @@ def bad_django_TestCase(runner):
         from hypothesis.extra.django._impl import HypothesisTestCase
 
         return not isinstance(runner, HypothesisTestCase)
+
+
+# _ATOMIC_TYPES was introduced as an optimization in 3.12's dataclasses.
+_ATOMIC_TYPES = frozenset(
+    {
+        types.NoneType,
```

Comment:
I thought I was being clever by backporting this optimization, but `types.NoneType` is new in python3.10, and I don't particularly want to special case that for an optimization that wasn't even present in older versions to begin with.  It's probably more effort than it's worth to try and muck about with this. If people want the micro performance improvement, I think it's reasonable to say they should upgrade their python version.